File: argon-backend-jvm/src/main/scala/dev/argon/backend/jvm/jdkloader/JavaLibrary.scala
package dev.argon.backend.jvm.jdkloader

import java.io.InputStream
import java.nio.file.Path

import dev.argon.compiler.ErrorList
import dev.argon.compiler.core.ModuleId
import zio._
import zio.blocking.Blocking


sealed trait JavaLibrary {

  def findClass(pkg: String, name: String): ZManaged[Blocking, Option[ErrorList], InputStream]
}

abstract class JavaModule extends JavaLibrary {
  val moduleName: String
}

abstract class ClassPath extends JavaLibrary

object JavaLibrary {

  def idFromModuleName(name: String): ModuleId = ModuleId("java:" + name)
  val classpathModuleId: ModuleId = ModuleId("java!classpath")

}

File: argon-backend-jvm/src/main/java/dev/argon/backend/jvm/jdkloader/JavaLibrary.java
package dev.argon.backend.jvm.jdkloader;

import dev.argon.compiler.core.ModuleId;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.io.InputStream;

public interface JavaLibrary {
    @Nullable InputStream findClass(@NotNull String pkg, @NotNull String name) throws IOException;

    <T> T visit(@NotNull JavaLibraryVisitor<T> visitor);


    public static @NotNull ModuleId idFromModuleName(@NotNull String name) {
        return new ModuleId("java:" +name);
    }

    public static final ModuleId classpathModuleId = new ModuleId("java!classpath");
}
